### ISSUE TITLE
Add partial conversion functions for converting between types

### ISSUE THREAD

**Issue by ChShersh** [kowainik/relude#114]
Especially useful for numeric types:

```haskell
integerToNatural :: Integer -> Maybe Natural
integerToInt     :: Integer -> Maybe Int
intToWord        :: Int -> Maybe Word
wordToInt        :: Word -> Maybe Int
```

Maybe there could be some abstraction for such kind of things that can looks like this:

```haskell
class PartialProjection small big where
    widen   :: small -> big
    shorten :: big -> Maybe small
```

Or even generalise this construction to be able to have `Maybe` or not have it in both directions.

Looks like `Prism` but dunno. To be discussed.

**Comment by cronokirby** [comment]
I like the idea of the type class personally, seems like a useful abstraction to have, though I'm not sure if we want to extend it beyond just numeric types.
Something like:
```haskell
class Embeded small big where
    expand :: small -> big
    contract :: big -> Maybe small
```
conveys the idea of the numeric types we want to abstract over, without generalising too far.

I think this does generalise to some kind of prism more generally, but I think there's a use case for a less general abstraction, since this is indeed quite useful.

**Comment by ChShersh** [comment]
@cronokirby Yes, this idea is worth exploring 👍 Maybe it call it `Expandable` or something like that. I'm okay giving it a try as one of the `Relude.Extra.*` modules. The most difficult part here is naming. Name `expand` appears too often. Maybe can think of something better 🤔 

Also, in general it's usually better to have type with `Either` to tell the reason of failure, so the following is better:

```haskell
contract :: big -> Either String small
```

And it's quite trivial to write `optionalContract :: big -> Maybe small`.

Rather than numeric hierarchy I see the following instance that might be useful:

```haskell
instance Expandable (Maybe a)    (Either e a)  -- this is `rightToMaybe` from `relude`
instance Expandable (NonEmpty a) [a]
```

And the law is:

```haskell
contract . expand ≡ pure
```

If we're interested in having only numeric things then in most cases it's enough to have the following function implemented by @jiegillet in https://github.com/kowainik/tomland/pull/120

```haskell
integerToBounded :: forall a. (Integral a, Bounded a) => Integer -> Maybe a
```

> Btw, I just noticed that similar function is already in `base`:
> * https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Bits.html#v:toIntegralSized

Though, it doesn't work with `Integer -> Maybe Natural`... 😞 

Another interesting idea similar to such abstraction is already implemented for injections by @int-index in his `inj` and `inj-base` packages. The code is very interesting by itself:

* http://hackage.haskell.org/package/inj-base-0.2.0.0/docs/Inj-Base.html

**Comment by cronokirby** [comment]
I'm not a big fan of having `String` as an error type, but an `Embeded` or some other similar typeclass to generalise this idea could be interesting to explore.

I think that for now we could add `integerToBounded` to `relude`, although I'm not sure where the best fit would be, but I do think it warrants addition because it seems consistently useful enough.

**Comment by ChShersh** [comment]
Yeah, I also don't know where to put `integerToBounded`... It could be `Relude.Extra.Enum` if only it was called `Relude.Extra.BoundedEnum`. Or if we had `Relude.Numeric` (probably a good idea, should be discussed under separate issue).

Obviously, `String` is not a good type for error message. But it's hard to have open sum type with ability for users to add more constructors. We can only have `newtype ExpandableError = ExpandableError Text` and it's better than `String`, but still...

**Comment by ChShersh** [comment]
Partial numeric functions are implemented in #128.
There's `witherable` for partial mappings but see #113.
I think experiments like `Expandable` are too new and should be explored in separate packages first before adding to custom prelude. So I think that this issue can be closed.

**Comment by tysonzero** [comment]
So it seems to me like the fundamental underlying concept is "there is a natural way to convert from a to b", where these conversions are sometimes partial, sometimes injective and sometimes surjective. From this the most fine grained hierarchy we can build is the following:

```
class PCast a b where
    pCast :: a -> Maybe b

class PCast a b => PSurj a b where
     pSurj :: a -> Maybe b

class (PCast a b, PInj b a) => PInj a b where
    pInj :: a -> Maybe b

class (PSurj a b, PInj a b, Inj b a) => PBij a b where
    pBij :: a -> Maybe b

class PCast a b => Cast a b where
    cast :: a -> b

class (PSurj a b, Cast a b) => Surj a b where
    surj :: a -> b

class (PInj a b, Cast a b, PBij b a) => Inj a b where
    inj :: a -> b

class (Surj a b, Inj a b, Bij b a) => Bij a b where
    bij :: a -> b
```

Quick note that PSurj and PBij don't quite correspond to the mathematical definition of a partial surjection/bijection, as we are referring to the entire output type as opposed to just the image of the function. For example mathematically `f :: Bool -> Bool`, `f False = False` is a partial bijection, but above I consider it to be a partial injection and not surjective.

The laws for this hierarchy should be fairly clear from context, namely that superclass functions in the same direction should return the same thing as the original function (modulo adding a Just), and that superclass functions in the opposite directions should totally invert the original function.

This hierarchy may seem overkill but it's worth noting there are practical examples of pairs of types for every above class that match that class and its superclasses (by definition) but none of its subclasses:

```
PCast String Regex
PSurj String Int
PInj Int Nat
PBij Integer Int
Cast String Parser
Surj UTCTime Day
Inj Int Integer
Bij Int Int
```

The `String` `Regex` one is a bit cheeky as it's assuming the `Regex` type internally uses functions and thus can't be deeply inspected, and also assumes that some combinators exist for the `Regex` that can't be expressed in string form. But if you drop `PCast` the hierarchy gets fractured, so once you commit to the rest you kind of have to keep it.

I do not currently see a library that implements anything similar to the above, so I honestly might implement such a library as sometimes I know there is a single intuitive way to go from `A` to `B` and want GHC to do the work for me. The `inj` library looks interesting although the extensive non-Maybe partiality more or less turns me away from it.
